src/reason/core.cljs

```diff
@@ -42,17 +42,30 @@
   (if (empty? match-rule)
     (constantly false)
     (fn [record]
-      (let [key (key-for-prefix key-prefix record)
+      (let [key (key-for-prefix key-prefix (keys record))
             val (get record key)]
         (match? match-rule val)))))
 
+(defn ^:private create-pred
+  "Given keys and the subrules, create a predicate.
+  If not all prefixes are in provided kys vector then predicates return false."
+  [kys subrules]
+  (let [key-prefixes (map :key-prefix subrules)
+        not-all-prefixes-exist?
+        (not-every? #(some? (key-for-prefix % kys)) key-prefixes)
+        pred (if (and (seq kys) not-all-prefixes-exist?)
+               #(constantly false)
+               #(parsed-subrule->pred %))]
+    (map (fn [subrule] (assoc subrule :pred (pred subrule))) subrules)))
+
 (defn rule->pred
-  "Given a rule, give a predicate for that rule."
-  [rule]
+  "Given a rule, give a predicate for that rule.
+   Returns nil if the rule contains keys not in provided keys param."
+  [rule & [rule-keys]]
```

Comment:
Style nit: this should be a second arity, as per http://clojure-doc.org/articles/language/functions.html#multi-arity-functions